com.unity.template.renderstreaming-hd/Assets/Scripts/RenderStreaming.cs

```diff
@@ -152,38 +193,35 @@ void OnDisable()
             {
                 this.m_signaling.Stop();
                 this.m_signaling = null;
+                this.m_negotiationneededCounter = 0;
             }
         }
 
         void OnOffer(ISignaling signaling, DescData e)
         {
-            RTCSessionDescription _desc;
-            _desc.type = RTCSdpType.Offer;
-            _desc.sdp = e.sdp;
             var connectionId = e.connectionId;
             if (m_mapConnectionIdAndPeer.ContainsKey(connectionId))
             {
-                return;
+                Debug.LogError($"connection:{connectionId} peerConnection already exist");
             }
-            var pc = new RTCPeerConnection();
-            m_mapConnectionIdAndPeer.Add(e.connectionId, pc);
 
-            pc.OnDataChannel = new DelegateOnDataChannel(channel => { OnDataChannel(pc, channel); });
-            pc.SetConfiguration(ref m_conf);
-            pc.OnIceCandidate = new DelegateOnIceCandidate(candidate =>
+            var pc = PrepareNewPeerConnection(signaling, connectionId, false);
+
+            RTCSessionDescription _desc;
+            _desc.type = RTCSdpType.Offer;
+            _desc.sdp = e.sdp;
+
+            var opRemoteDesc = pc.SetRemoteDescription(ref _desc);
+            while (opRemoteDesc.MoveNext())
```

Comment:
Isn't this blocking the main thread?

Reply:
As far as I can see, the main thread is not blocked because it's running from another thread.
(Ithink, `OnOffer` Invoked by thread about managed by signaling.)

Reply:
I see. 
In my opinion, the signaling instance invokes a delegate on another thread is hard to understand.

Reply:
That's true...
Change method IEnumerutor and invoke corutine from signaling instance.
It should probably work fine.

Reply:
I fixed